syntax = "proto3";

package aura.negotiation.v1;

// Service for handling negotiations.
service NegotiationService {
    // RPC method for negotiating offers.
    rpc Negotiate (NegotiateRequest) returns (NegotiateResponse) {}

    rpc Search (SearchRequest) returns (SearchResponse);

    // RPC method for retrieving system status and metrics.
    rpc GetSystemStatus (GetSystemStatusRequest) returns (GetSystemStatusResponse);

    // RPC method for checking crypto payment status and retrieving secrets.
    rpc CheckDealStatus (CheckDealStatusRequest) returns (CheckDealStatusResponse);
}

message NegotiateRequest {
  string request_id = 1;      // Unique identifier for request idempotency.
  string item_id = 2;         // Identifier of the item being negotiated.
  double bid_amount = 3;      // The proposed bid amount.
  string currency_code = 4;   // Currency code, e.g., "USD".
  
  // Metadata for the agent identity.
  AgentIdentity agent = 5;
}

message AgentIdentity {
  string did = 1;             // Decentralized identifier.
  float reputation_score = 2; // Reputation score from middleware.
}

message NegotiateResponse {
  string session_token = 1;   // Token for the negotiation session.
  int64 valid_until_timestamp = 2; // Timestamp until the session is valid.

  // Polymorphic result: either acceptance, counteroffer, rejection, or UI request.
  oneof result {
    OfferAccepted accepted = 3;
    OfferCountered countered = 4;
    OfferRejected rejected = 5;
    JitUiRequest ui_required = 6;
  }
}

message OfferAccepted {
  double final_price = 1;

  // Either reservation_code (legacy) or crypto_payment (new)
  oneof reveal_method {
    string reservation_code = 2;
    CryptoPaymentInstructions crypto_payment = 3;
  }
}

// Instructions for completing a crypto payment to unlock the deal
message CryptoPaymentInstructions {
  string deal_id = 1;              // UUID of the locked deal
  string wallet_address = 2;       // Payment destination address
  double amount = 3;               // Amount to pay
  string currency = 4;             // "SOL" or "USDC"
  string memo = 5;                 // Unique memo to include in transaction
  string network = 6;              // "mainnet-beta", "devnet", "testnet"
  int64 expires_at = 7;            // Unix timestamp when deal expires
}

// Request to check the status of a locked deal
message CheckDealStatusRequest {
  string deal_id = 1;              // UUID of the deal to check
}

// Response containing deal status and secret (if paid)
message CheckDealStatusResponse {
  string status = 1;               // "PENDING", "PAID", "EXPIRED", "NOT_FOUND"

  // Only present if status == "PAID"
  DealSecret secret = 2;
  PaymentProof proof = 3;

  // Only present if status == "PENDING"
  CryptoPaymentInstructions payment_instructions = 4;
}

// The revealed secret after successful payment
message DealSecret {
  string reservation_code = 1;     // The actual reservation code
  string item_name = 2;            // Name of purchased item
  double final_price = 3;          // Final agreed price
  int64 paid_at = 4;               // Unix timestamp of payment confirmation
}

// Proof of on-chain payment
message PaymentProof {
  string transaction_hash = 1;     // Blockchain transaction ID
  string block_number = 2;         // Block number for verification
  string from_address = 3;         // Payer's wallet address
  int64 confirmed_at = 4;          // Unix timestamp of confirmation
}

message OfferCountered {
  double proposed_price = 1;  // The proposed counter price.
  string reason_code = 2;     // Machine-readable reason code, e.g., "BELOW_FLOOR_LIMIT".
  string human_message = 3;   // Human-readable message for display.
}

message OfferRejected {
  string reason_code = 1;     // Reason code for the rejection.
}

message JitUiRequest {
  string template_id = 1;     // Identifier for the UI template, e.g., "high_value_confirm".
  map<string, string> context_data = 2; // Data for rendering the UI.
}

message SearchRequest {
  string query = 1;        // "Cheap place in Bali"
  int32 limit = 2;         // Top N results
  double min_similarity = 3; // Отсечка мусора (0.0 - 1.0)
}

message SearchResponse {
  repeated SearchResultItem results = 1;
}

message SearchResultItem {
  string item_id = 1;
  string name = 2;
  double base_price = 3;
  double similarity_score = 4; // Насколько это соответствует запросу
  string description_snippet = 5; // Для контекста агенту
}

message GetSystemStatusRequest {}

message GetSystemStatusResponse {
  string status = 1;              // "ok" | "error"
  double cpu_usage_percent = 2;
  double memory_usage_mb = 3;
  string timestamp = 4;
  bool cached = 5;
}